File: geekbang/algorithm-models/7.4- Floyd 求最短路代码模板/findTheCity.cpp
// C/C++
// LeetCode 1334
class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        // 邻接矩阵初值：i到i长度为0，没有边长度为INF，其余为输入的边
        vector<vector<int>> d(n, vector<int>(n, 1e9));
        for (auto& edge : edges) {
            int x = edge[0], y = edge[1], z = edge[2];
            d[x][y] = d[y][x] = z;
        }
        for (int i = 0; i < n; i++) d[i][i] = 0;
        // Floyd算法
        for (int k = 0; k < n; k++)
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++)
                    d[i][j] = min(d[i][j], d[i][k] + d[k][j]);
        // 统计答案
        int ansCount = n, ans;
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = 0; j < n; j++)
                if (i != j && d[i][j] <= distanceThreshold) count++;
            if (count <= ansCount) {
                ansCount = count;
                ans = i;
            }
        }
        return ans;
    }
};

File: geekbang/algorithm-models/7.4- Floyd 求最短路代码模板/findTheCity.py
# Python
# LeetCode 1334
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        # 邻接矩阵初值：i到i长度为0，没有边长度为INF，其余为输入的边
        d = [[1e9] * n for i in range(n)]
        for edge in edges:
            x, y, z = edge[0], edge[1], edge[2]
            d[x][y] = d[y][x] = z
        for i in range(n):
            d[i][i] = 0
        # Floyd算法
        for k in range(n):
            for i in range(n):
                for j in range(n):
                    d[i][j] = min(d[i][j], d[i][k] + d[k][j])
        # 统计答案
        ansCount, ans = n, 0
        for i in range(n):
            count = 0
            for j in range(n):
                if i != j and d[i][j] <= distanceThreshold:
                    count += 1
            if count <= ansCount:
                ansCount = count
                ans = i
        return ans
